#166238097 implement Add a new book endpoint

obayomi96 opened the pull request:
#### What does this PR do?
This PR creates an endpoint for a user to add a new book.
Restructures the routes files in the routes directory
Add a new table to the buildTables and dropTables file as well as seeds data
Refactors the migration scripts

#### Description of Task to be completed?
Have the following endpoints working
> POST api/v1/books

#### How should this be manually tested?
> Clone the repo and cd into the directory
> Run npm install
> Run npm run Start-dev
> Use postman to test all endpoints above
> key: Content-Type value: application/json
navigate to the above url to test all endpoint

#### Any background context you want to provide?
Books form the backbone of our application and will introduce us to CRUD which is the building block of most web applications. This endpoint creates a book model to insert a book into the books table in the database. That model is then used to implement the logic of adding the book specific data and the response. 

#### What are the relevant pivotal tracker stories?
#166238097

#### Screenshots (if appropriate)

![Screenshot (238)](https://user-images.githubusercontent.com/43539944/58835238-5816be00-864d-11e9-911b-d4852533f37b.png)

ayodejiAA committed:
Commit to rebase

ayodejiAA committed:
Commit to rebase

ayodejiAA committed:
Commit to rebase

ayodejiAA committed:
Commit to rebase

ayodejiAA committed:
Merge with develop

ayodejiAA committed:
Merge with develop

ayodejiAA committed:
Merge with develop

ayodejiAA committed:
Merge with develop

ayodejiAA committed:
feature(authentication): User should receive JWT
 Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

[Finishes #166238094]

commit f34bfd7cda1293f06077387fca65e74905493a20
Merge: d97033f c3b79d5
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:59:23 2019 +0100

    Merge pull request #6 from los-deferal-02/ch-setup-testing-framework-166237755

    #166237755 Setup Automated Testing Framework

commit d97033f5a094c8c61c97845783627f668bb89a6a
Merge: 145cfd5 ff4b0e0
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:58:04 2019 +0100

    Merge pull request #8 from los-deferal-02/ch-deploy-to-heroku-166237764

    #166237764 Deploy base application to Heroku

commit c3b79d5624ba67cfc425c86abb11bd941057327b
Merge: 9223bf4 145cfd5
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Thu May 30 18:06:26 2019 +0100

    chore(test): fix merge conflict

commit ff4b0e019da81d51a9ad32752685df850c17fb27
Author: nerocodes <neropaulej@gmail.com>
Date:   Wed May 29 12:30:53 2019 +0100

    chore(heroku): added app.json for review apps
    [Finishes #166237764]

commit 9223bf480b3078ece556e934900e6c166482bdce
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Tue May 28 23:32:24 2019 +0100

    chore(testing): set up automated testing framework

    - used combination of mocha, chai, and chai-http
    - add test script to package.json file
    - write two test files for root and undefined routes

    [Finishes #166237755]

ayodejiAA committed:
feature(authentication): User should receive JWT
 Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

[Finishes #166238094]

commit f34bfd7cda1293f06077387fca65e74905493a20
Merge: d97033f c3b79d5
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:59:23 2019 +0100

    Merge pull request #6 from los-deferal-02/ch-setup-testing-framework-166237755

    #166237755 Setup Automated Testing Framework

commit d97033f5a094c8c61c97845783627f668bb89a6a
Merge: 145cfd5 ff4b0e0
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:58:04 2019 +0100

    Merge pull request #8 from los-deferal-02/ch-deploy-to-heroku-166237764

    #166237764 Deploy base application to Heroku

commit c3b79d5624ba67cfc425c86abb11bd941057327b
Merge: 9223bf4 145cfd5
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Thu May 30 18:06:26 2019 +0100

    chore(test): fix merge conflict

commit ff4b0e019da81d51a9ad32752685df850c17fb27
Author: nerocodes <neropaulej@gmail.com>
Date:   Wed May 29 12:30:53 2019 +0100

    chore(heroku): added app.json for review apps
    [Finishes #166237764]

commit 9223bf480b3078ece556e934900e6c166482bdce
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Tue May 28 23:32:24 2019 +0100

    chore(testing): set up automated testing framework

    - used combination of mocha, chai, and chai-http
    - add test script to package.json file
    - write two test files for root and undefined routes

    [Finishes #166237755]

ayodejiAA committed:
feature(authentication): User should receive JWT
 Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

[Finishes #166238094]

commit f34bfd7cda1293f06077387fca65e74905493a20
Merge: d97033f c3b79d5
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:59:23 2019 +0100

    Merge pull request #6 from los-deferal-02/ch-setup-testing-framework-166237755

    #166237755 Setup Automated Testing Framework

commit d97033f5a094c8c61c97845783627f668bb89a6a
Merge: 145cfd5 ff4b0e0
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:58:04 2019 +0100

    Merge pull request #8 from los-deferal-02/ch-deploy-to-heroku-166237764

    #166237764 Deploy base application to Heroku

commit c3b79d5624ba67cfc425c86abb11bd941057327b
Merge: 9223bf4 145cfd5
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Thu May 30 18:06:26 2019 +0100

    chore(test): fix merge conflict

commit ff4b0e019da81d51a9ad32752685df850c17fb27
Author: nerocodes <neropaulej@gmail.com>
Date:   Wed May 29 12:30:53 2019 +0100

    chore(heroku): added app.json for review apps
    [Finishes #166237764]

commit 9223bf480b3078ece556e934900e6c166482bdce
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Tue May 28 23:32:24 2019 +0100

    chore(testing): set up automated testing framework

    - used combination of mocha, chai, and chai-http
    - add test script to package.json file
    - write two test files for root and undefined routes

    [Finishes #166237755]

ayodejiAA committed:
feature(authentication): User should receive JWT
 Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

[Finishes #166238094]

commit f34bfd7cda1293f06077387fca65e74905493a20
Merge: d97033f c3b79d5
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:59:23 2019 +0100

    Merge pull request #6 from los-deferal-02/ch-setup-testing-framework-166237755

    #166237755 Setup Automated Testing Framework

commit d97033f5a094c8c61c97845783627f668bb89a6a
Merge: 145cfd5 ff4b0e0
Author: obije <37974813+akpante3@users.noreply.github.com>
Date:   Thu May 30 18:58:04 2019 +0100

    Merge pull request #8 from los-deferal-02/ch-deploy-to-heroku-166237764

    #166237764 Deploy base application to Heroku

commit c3b79d5624ba67cfc425c86abb11bd941057327b
Merge: 9223bf4 145cfd5
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Thu May 30 18:06:26 2019 +0100

    chore(test): fix merge conflict

commit ff4b0e019da81d51a9ad32752685df850c17fb27
Author: nerocodes <neropaulej@gmail.com>
Date:   Wed May 29 12:30:53 2019 +0100

    chore(heroku): added app.json for review apps
    [Finishes #166237764]

commit 9223bf480b3078ece556e934900e6c166482bdce
Author: Abdulfatai Jimoh <abdulfatai360@gmail.com>
Date:   Tue May 28 23:32:24 2019 +0100

    chore(testing): set up automated testing framework

    - used combination of mocha, chai, and chai-http
    - add test script to package.json file
    - write two test files for root and undefined routes

    [Finishes #166237755]

ayodejiAA committed:
feature(authentication): User should receive JWT
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
feature(authentication): User should receive JWT
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
feature(authentication): User should receive JWT
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
feature(authentication): User should receive JWT
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
Merge branch 'ft-implement-jwt-authentication-166238094' of https://github.com/los-deferal-02/Hello-Books into ft-implement-jwt-authentication-166238094

abdulfatai360 committed:
feature(email): create email module

[Delivers #166238095]

abdulfatai360 committed:
feature(email): create email module

[Delivers #166238095]

ayodejiAA committed:
Merge branch 'ft-implement-jwt-authentication-166238094' of https://github.com/los-deferal-02/Hello-Books into ft-implement-jwt-authentication-166238094

tolulope-od committed:
feature(auth):registration validation

Add input validation middleware
Create server response module

[Start #166238096]

chialuka committed:
pull controllers

obayomi96 committed:
feature(Add new book):implement
User can add a new book to the books table
Revamped the routes directory and decentralized route files
Created table and seed for migrations [#delivers #166238097]

obayomi96 committed:
feature(Add new book):implement
User can add a new book to the books table
Revamped the routes directory and decentralized route files
Created table and seed for migrations [#delivers #166238097]

obayomi96 committed:
feature(Add new book):implement
User can add a new book to the books table
Revamped the routes directory and decentralized route files
Created table and seed for migrations [#delivers #166238097]

obayomi96 committed:
feature(Add new book):implement
User can add a new book to the books table
Revamped the routes directory and decentralized route files
Created table and seed for migrations [#delivers #166238097]

chiboycalix commented on server/config/buildTables.js:
```diff
@@ -1,14 +1,26 @@
+    id SERIAL PRIMARY KEY,
+    title VARCHAR(100) UNIQUE NOT NULL,
+    body VARCHAR(100) NOT NULL,
+    description VARCHAR(100) NOT NULL,
```
Good job @obayomi96 , what's the reason behind making the description field to accept just 100 characters? won't it be better if it can accept more than that since it is a book's description? or better still you can make a TEXT field.

obayomi96 replied:
Thanks for pointing that out. I will rectify it.

chiboycalix replied:
cool

chiboycalix commented on server/test/login.spec.js:
```diff
@@ -0,0 +1,87 @@
+    });
+
+    it('500 internal error if server encounters error', (done) => {
+      const stub = sinon.stub(pool, 'query')
```
Please can you explain to me what this line of code does? i'll like to know

obayomi96 replied:
That was implemented by @Xwebyna . But you can read mre about sinon and stub online. I think it pauses  the initial function and run its own which is the query.

chialuka replied:
Sinon was used here to mock an API request. Since you can't generate a 500 server error to test, Sinon is used to do it. That way, your test coverage can cover server errors.

obayomi96 replied:
Thanks for clarifying that @LukasChiama

chiboycalix replied:
Thank you @LukasChiama I understand now

obayomi96 committed:
fix- Set description to TEXT in books table

obayomi96 committed:
fix- Set description to TEXT in books table

akpante3 commented on server/config/dropTables.js:
```diff
@@ -1,16 +1,18 @@
-    await pool.query('DROP TABLE IF EXISTS USERS');
-    return 'Tables dropped successfully';
+    await pool.query('DROP TABLE IF EXISTS users, books');
+    log('Tables dropped successfully');
```
remove logs

obayomi96 replied:
Done

akpante3 commented on server/controllers/userController.js:
```diff
@@ -0,0 +1,77 @@
+    try {
+      const data = req.body;
+      const { password } = data;
+      data.password = await encryptPassword(password);
```
why distructure data on `line 25` and still make use of `data.password` on` line 26`

obayomi96 replied:
This was implemented in the branch I pulled to work on. But password was destructured from data so that it can be encrypted before being sent into the db.

akpante3 commented on server/controllers/userController.js:
```diff
@@ -0,0 +1,77 @@
+      // Login with username or email address
+      const user = await userModel.findUserInput(userLogin);
+      if (!user) {
+        return res.status(404).json({
```
API response should be extracted into a response file.

tolulope-od replied:
@akpante3 I was assigned the task to create a response module and there's an open PR that implements it in #16

obayomi96 replied:
Yes, this was implemented by @Xwebyna .. I noticed some stories on the PT has to be done together with some other stories.

chialuka committed:
feature(verify email):add email verification
[Delivers #166238090]

chialuka committed:
fix merge conflicts

chialuka committed:
replace logs with debug

chialuka committed:
add API documentation

abdulfatai360 committed:
feature(email): update local branch with develop

abdulfatai360 committed:
feature(email): create module for sending emails

tolulope-od committed:
feature(auth):login validation

Add validation for user login
Add more test cases
Refactor db column names

[Finishes #166238096]

chialuka commented on hello-books-swagger.json:
```diff
@@ -1,97 +1,200 @@
+				}
+			}
+		}
+	},
```
Great job bro. This is well done. 

The team agreed that for every endpoint added, the documentation will also be added. Could you work on adding that?

obayomi96 replied:
Thanks for the reminder. That has been done.

tolulope-od committed:
feature(auth):resolve merge conflicts

Add morgan for logging in development mode

[Finishes #166238096]

ayodejiAA committed:
feature(authentication): User should receive JWT
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

obayomi96 committed:
Merge branch 'ft-user-can-add-book-166238097' of https://github.com/los-deferal-02/Hello-Books into ft-user-can-add-book-166238097

ayodejiAA committed:
Fix test issue

chialuka reviewed (commented):
Great job @obayomi96. Could you also add the API documentation for the endpoint?

obayomi96 committed:
Swagger documention for addBook endpoint

obayomi96 committed:
Swagger documention for addBook endpoint

obayomi96 committed:
fix conflicts

akpante3 committed:
Merge pull request #16 from los-deferal-02/ft-user-registration-validation-166238096

#166238096 Users should receive descriptive signup/Registration validation errors

chialuka committed:
fix merge conflicts and scripts

akpante3 committed:
Merge pull request #15 from los-deferal-02/ft-verify-email-166238090

#166238090 Build Email Verification Service

ayodejiAA committed:
feature(authentication): User should receive JWT---
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

ayodejiAA committed:
feature(authentication): User should receive JWT..:
Upon successful registration/login
- Write tests for the feature
- Document feature with swagger

akpante3 committed:
Merge pull request #11 from los-deferal-02/ft-implement-jwt-authentication-166238094

#166238094 Users should receive a JWT on Successful Registration/Login

abdulfatai360 committed:
Merge branch 'develop' into ft-create-email-module-166238095

akpante3 committed:
Merge pull request #12 from los-deferal-02/ft-create-email-module-166238095

#166238095 Create Mail Module for Sending Email Messages

Nerocodes committed:
feature(api): added view, edit and create user profile endpoints
[Delivers #166238072]

akpante3 committed:
Merge pull request #17 from los-deferal-02/ft-user-profile-166238072

#166238072 Users should be able to create their profiles

ayodejiAA committed:
feature(signup): Users can register as patron or author

ayodejiAA committed:
feature(signup): Users can register as patron or author

[Finishes #166608419]

akpante3 committed:
Merge pull request #19 from los-deferal-02/ft-user-signup-as-author-166608419

#166608419 feature(signup): Users can register as patron or author

tolulope-od committed:
chore(merge):resolve conflicts

tolulope-od committed:
Add emailModule file from develop

tolulope-od committed:
chore(health):resolve merge conflicts

Resolve merge conflicts
Refactor routing logic
Fix tests

[Finishes #166238097]

tolulope-od committed:
chore(test):fix test coverage
